File: src/jdwp/protocol.ts
// JDWP packet framing and the command/event constants cappu's debugger uses.
// A JDWP packet is an 11-byte header followed by a command-set-specific body:
//   length:u4  id:u4  flags:u1  then (commandSet:u1 command:u1) for a command
//                                 or (errorCode:u2)              for a reply.
// All multi-byte values are big-endian. Replies set the 0x80 flag bit and carry
// the id of the command they answer; events arrive as Event.Composite commands.
//
// Port reference for togo/internal/jdwp/protocol.go.

export const HEADER_LEN = 11;
export const FLAG_REPLY = 0x80;
export const HANDSHAKE = "JDWP-Handshake";

export type Packet =
  | { kind: "command"; id: number; commandSet: number; command: number; data: Buffer }
  | { kind: "reply"; id: number; errorCode: number; data: Buffer };

export function encodeCommandPacket(
  id: number,
  commandSet: number,
  command: number,
  data: Buffer,
): Buffer {
  const header = Buffer.allocUnsafe(HEADER_LEN);
  header.writeUInt32BE(HEADER_LEN + data.length, 0);
  header.writeUInt32BE(id >>> 0, 4);
  header.writeUInt8(0, 8); // flags: command packet
  header.writeUInt8(commandSet, 9);
  header.writeUInt8(command, 10);
  return Buffer.concat([header, data]);
}

/** Decode exactly one full packet (header + body); caller frames by length. */
export function decodePacket(buf: Buffer): Packet {
  const id = buf.readUInt32BE(4);
  const flags = buf.readUInt8(8);
  const data = buf.subarray(HEADER_LEN);
  if (flags & FLAG_REPLY) {
    return { kind: "reply", id, errorCode: buf.readUInt16BE(9), data };
  }
  return {
    kind: "command",
    id,
    commandSet: buf.readUInt8(9),
    command: buf.readUInt8(10),
    data,
  };
}

/**
 * Pull the first complete packet off a stream buffer. Returns the decoded
 * packet plus the unconsumed remainder, or null when fewer than a full packet's
 * bytes have arrived yet (the length prefix tells us how many we need).
 */
export function tryReadPacket(buf: Buffer): { packet: Packet; rest: Buffer } | null {
  if (buf.length < HEADER_LEN) return null;
  const length = buf.readUInt32BE(0);
  if (buf.length < length) return null;
  return { packet: decodePacket(buf.subarray(0, length)), rest: buf.subarray(length) };
}

// Command sets and the commands within them that cappu issues. Numbers are from
// the JDWP spec (java.se/docs jdwp-protocol). Only the subset the debugger uses
// is listed.
export const CommandSet = {
  VirtualMachine: 1,
  ReferenceType: 2,
  ClassType: 3,
  Method: 6,
  ObjectReference: 9,
  StringReference: 10,
  ThreadReference: 11,
  ArrayReference: 13,
  EventRequest: 15,
  StackFrame: 16,
  Event: 64,
} as const;

export const VirtualMachineCmd = {
  Version: 1,
  ClassesBySignature: 2,
  AllClasses: 3,
  AllThreads: 4,
  Dispose: 6,
  IDSizes: 7,
  Suspend: 8,
  Resume: 9,
  Exit: 10,
  CreateString: 11,
} as const;

export const ReferenceTypeCmd = {
  Signature: 1,
  Fields: 4,
  Methods: 5,
  SourceFile: 7,
} as const;

export const MethodCmd = {
  LineTable: 1,
  VariableTable: 2,
  VariableTableWithGeneric: 5,
} as const;

export const ObjectReferenceCmd = {
  ReferenceType: 1,
  GetValues: 2,
  InvokeMethod: 6,
} as const;

export const StringReferenceCmd = { Value: 1 } as const;

export const ThreadReferenceCmd = {
  Name: 1,
  Suspend: 2,
  Resume: 3,
  Status: 4,
  Frames: 6,
  FrameCount: 7,
} as const;

export const ArrayReferenceCmd = { Length: 1, GetValues: 2 } as const;

export const EventRequestCmd = { Set: 1, Clear: 2 } as const;

export const StackFrameCmd = { GetValues: 1 } as const;

export const EventCmd = { Composite: 100 } as const;

// JDWP enums used when building event requests and decoding events/values.
export const EventKind = {
  SINGLE_STEP: 1,
  BREAKPOINT: 2,
  THREAD_START: 6,
  THREAD_DEATH: 7,
  CLASS_PREPARE: 8,
  VM_START: 90,
  VM_DEATH: 99,
} as const;

export const SuspendPolicy = { NONE: 0, EVENT_THREAD: 1, ALL: 2 } as const;

export const TypeTag = { CLASS: 1, INTERFACE: 2, ARRAY: 3 } as const;

export const ModifierKind = {
  Count: 1,
  ClassMatch: 5,
  LocationOnly: 7,
  Step: 10,
} as const;

export const StepSize = { MIN: 0, LINE: 1 } as const;
export const StepDepth = { INTO: 0, OVER: 1, OUT: 2 } as const;

// Value tag bytes (JDWP Value): the leading tag of a tagged value tells you how
// to decode the rest. Only the subset the variable formatter handles in v1.
export const Tag = {
  ARRAY: 91, // '['
  BYTE: 66, // 'B'
  CHAR: 67, // 'C'
  OBJECT: 76, // 'L'
  FLOAT: 70, // 'F'
  DOUBLE: 68, // 'D'
  INT: 73, // 'I'
  LONG: 74, // 'J'
  SHORT: 83, // 'S'
  VOID: 86, // 'V'
  BOOLEAN: 90, // 'Z'
  STRING: 115, // 's'
  THREAD: 116, // 't'
} as const;

File: togo/internal/jdwp/protocol.go
package jdwp

// JDWP packet framing and the command/event constants cappu's debugger uses. A
// JDWP packet is an 11-byte header followed by a command- or reply-specific
// body; all multi-byte values are big-endian. Replies set the 0x80 flag and
// carry the id of the command they answer; events arrive as Event.Composite.
// Port of src/jdwp/protocol.ts.

import "encoding/binary"

const (
	HeaderLen = 11
	FlagReply = 0x80
	Handshake = "JDWP-Handshake"
)

// Packet is a decoded JDWP packet, either a reply or a command (event).
type Packet struct {
	IsReply    bool
	ID         uint32
	CommandSet byte
	Command    byte
	ErrorCode  uint16
	Data       []byte
}

// EncodeCommandPacket frames a command packet (header + body).
func EncodeCommandPacket(id uint32, set, cmd byte, data []byte) []byte {
	out := make([]byte, HeaderLen+len(data))
	binary.BigEndian.PutUint32(out[0:], uint32(HeaderLen+len(data)))
	binary.BigEndian.PutUint32(out[4:], id)
	out[8] = 0 // flags: command packet
	out[9] = set
	out[10] = cmd
	copy(out[HeaderLen:], data)
	return out
}

// DecodePacket decodes exactly one full packet (header + body).
func DecodePacket(b []byte) Packet {
	p := Packet{
		ID:   binary.BigEndian.Uint32(b[4:]),
		Data: b[HeaderLen:],
	}
	if b[8]&FlagReply != 0 {
		p.IsReply = true
		p.ErrorCode = binary.BigEndian.Uint16(b[9:])
	} else {
		p.CommandSet = b[9]
		p.Command = b[10]
	}
	return p
}

// TryReadPacket pulls the first complete packet off a stream buffer, returning
// it plus the unconsumed remainder; ok is false when a full packet has not
// arrived yet.
func TryReadPacket(buf []byte) (packet Packet, rest []byte, ok bool) {
	if len(buf) < HeaderLen {
		return Packet{}, buf, false
	}
	length := int(binary.BigEndian.Uint32(buf[0:]))
	if len(buf) < length {
		return Packet{}, buf, false
	}
	return DecodePacket(buf[:length]), buf[length:], true
}

// Command sets.
const (
	CSVirtualMachine  byte = 1
	CSReferenceType   byte = 2
	CSClassType       byte = 3
	CSMethod          byte = 6
	CSObjectReference byte = 9
	CSStringReference byte = 10
	CSThreadReference byte = 11
	CSArrayReference  byte = 13
	CSEventRequest    byte = 15
	CSStackFrame      byte = 16
	CSEvent           byte = 64
)

// VirtualMachine commands.
const (
	VMVersion            byte = 1
	VMClassesBySignature byte = 2
	VMAllClasses         byte = 3
	VMAllThreads         byte = 4
	VMDispose            byte = 6
	VMIDSizes            byte = 7
	VMSuspend            byte = 8
	VMResume             byte = 9
	VMExit               byte = 10
	VMCreateString       byte = 11
)

// ReferenceType commands.
const (
	RTSignature  byte = 1
	RTFields     byte = 4
	RTMethods    byte = 5
	RTSourceFile byte = 7
)

// Method commands.
const (
	MLineTable     byte = 1
	MVariableTable byte = 2
)

// ObjectReference commands.
const (
	ORReferenceType byte = 1
	ORGetValues     byte = 2
	ORInvokeMethod  byte = 6
)

// StringReference commands.
const (
	SRValue byte = 1
)

// ThreadReference commands.
const (
	TRName       byte = 1
	TRSuspend    byte = 2
	TRResume     byte = 3
	TRStatus     byte = 4
	TRFrames     byte = 6
	TRFrameCount byte = 7
)

// ArrayReference commands.
const (
	ARLength    byte = 1
	ARGetValues byte = 2
)

// EventRequest commands.
const (
	ERSet   byte = 1
	ERClear byte = 2
)

// StackFrame commands.
const (
	SFGetValues byte = 1
)

// Event commands.
const (
	EVComposite byte = 100
)

// Event kinds.
const (
	EKSingleStep   byte = 1
	EKBreakpoint   byte = 2
	EKThreadStart  byte = 6
	EKThreadDeath  byte = 7
	EKClassPrepare byte = 8
	EKVMStart      byte = 90
	EKVMDeath      byte = 99
)

// Suspend policies.
const (
	SuspendNone        byte = 0
	SuspendEventThread byte = 1
	SuspendAll         byte = 2
)

// Type tags.
const (
	TypeTagClass     byte = 1
	TypeTagInterface byte = 2
	TypeTagArray     byte = 3
)

// EventRequest modifier kinds.
const (
	ModCount        byte = 1
	ModClassMatch   byte = 5
	ModLocationOnly byte = 7
	ModStep         byte = 10
)

// Step sizes and depths.
const (
	StepSizeMin   int32 = 0
	StepSizeLine  int32 = 1
	StepDepthIn   int32 = 0
	StepDepthOut  int32 = 2
	StepDepthOver int32 = 1
)

// JDWP value tag bytes.
const (
	TagArray   byte = 91  // '['
	TagByte    byte = 66  // 'B'
	TagChar    byte = 67  // 'C'
	TagObject  byte = 76  // 'L'
	TagFloat   byte = 70  // 'F'
	TagDouble  byte = 68  // 'D'
	TagInt     byte = 73  // 'I'
	TagLong    byte = 74  // 'J'
	TagShort   byte = 83  // 'S'
	TagVoid    byte = 86  // 'V'
	TagBoolean byte = 90  // 'Z'
	TagString  byte = 115 // 's'
	TagThread  byte = 116 // 't'
)
